interfaces/builtin: add dbus interface

jdstrand opened the pull request:
This implements dbus in support of 'Apps can't own session bus names'. Specifically, snaps may specify a slot using this interface which grants access to bind. Eg:

```
name: gnome-logs-udt
slots:
  org.gnome.Logs:
    interface: dbus
    bus: session
    name: org.gnome.Logs
```

This will allow the snap to own the well-known bus name 'org.gnome.Logs' and also integrate into the classic environment.

This also implements the plugs side and allows other snaps to connect to the interface with something like:

```
name: foo
plugs:
  org.gnome.Logs:
    interface: dbus
    bus: session
    name: org.gnome.Logs
```

Upon snap connect, this allows foo to introspect gnome-logs-udt and use its entire DBus API.

Please see https://github.com/snapcore/snapd/pull/1613#issuecomment-261350004 for complete details.

LP: #1590679

mvo5 commented on interfaces/builtin/dbus_app.go:
```diff
@@ -0,0 +1,280 @@
+	}
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		var snippets bytes.Buffer
```
(nitpick) You could write `snippets := bytes.NewBufferString("")` here to save 1 line.

jdstrand replied:
Done

mvo5 commented on interfaces/builtin/dbus_app.go:
```diff
@@ -0,0 +1,280 @@
+}
+
+// verify that name for bus is in list
+func (iface *DbusAppInterface) verifyNameInAttributes(bus string, name string, attribs map[string]interface{}) bool {
```
Am I overlooking something or is this function not used?

jdstrand replied:
No, you are right. This was used in the dbus-bind exploratory PR for making sure the slot side had an entry for the specified plug side. Removed. Thanks!

mvo5 commented on interfaces/builtin/dbus_app.go:
```diff
@@ -0,0 +1,280 @@
+}
+
+func (iface *DbusAppInterface) AutoConnect() bool {
+	return false
```
Maybe this just shows my ignorance, but earlier we have ` +Auto-Connect: yes ` in the docs and false here. Or is this a different auto-connect?

jdstrand replied:
Since we are implementing this in the permanent slot side it is auto-generated. There is no plugs side atm. I supposed it would be clearer if I just removed this.

mvo5 commented on interfaces/builtin/dbus_app_test.go:
```diff
@@ -0,0 +1,467 @@
+	c.Assert(err, IsNil)
+}
+
+/*
```
Maybe a comment here why it is commented out?

jdstrand replied:
Oh, this was meant to be removed-- it's a test for the plug side of the connection which we don't support in this PR (perhaps in the future as per description in this PR).

mvo5 commented:
Looks good, thanks for doing this work! Some questions inline but no blockers.

jdstrand commented:
@mvo5 I think I addressed all your questions. Thanks for the review! :)

mvo5 commented on interfaces/builtin/dbus_app.go:
```diff
@@ -0,0 +1,263 @@
+	return err
+}
+
+// Since we only implement the permanent slot side, this is meaningless but
```
Thanks for this comment!

mvo5 commented:
Nice! Thank you, :+1:

niemeyer commented:
@jdstrand I still feel like there's something not quite right with this interface. The whole logic seems to focus on the slot side only. What happens if someone connects to that slot?  How do we even hint to the user what can be connected to what?  How do we prevent an app from binding? If the answers are respectively "nothing", "we don't know", and "we can't", then something is wrong.

Taking our lesson from our conversation this week, isn't this actually more correct as dbus-setup, where the settings are done on the plug side, and only when actually granted via a connection?

niemeyer commented:
(s/dbus-setup/dbus-support)

jdstrand commented:
@niemeyer - first off, just want to reiterate this is an exploratory PR and that I'm happy to change direction if we determine that is best. I'm going to advocate in favor of this PR here (trying to be objective :) so that its merits are considered, but like I said, I'm happy to change direction if we decide that's what's best.

"@jdstrand I still feel like there's something not quite right with this interface. The whole logic seems to focus on the slot side only."

I focused on the slot side because slots provide an implementation of an interface and are typically services for other things to consume. In that light, this should be a slot because the app that performs a dbus bind is doing so for a reason-- so other applications can use whatever dbus API that the application provides. In the proposed implementation, we allow unconfined processes to interact with the rhythmbox dbus API and this should work well for classic systems. This PR does not deal with the plugs side however it is clear that this is possible in future iterations-- eg, rhythmbox might use dbus-app to bind to org.gnome.Rhythmbox and then a rhythmbox-remote app could plug into it and use the rhythmbox dbus API provided by the rhythmbox snap. While there are open questions, it seems natural.

Admittedly, we could turn this around and call it 'dbus-bind-support' or 'freedesktop-app-support' and implement it as a plug, but then we won't be able to conveniently extend the interface to connect a snap to another that plugs its interface (eg, rhythmbox with rhythmbox-remote, respectively).

The proposed PR's upside is that I think it more naturally reflects what the dbus bind is intending to do, and it allows us to iterate on it so we can allow arbitrary snap connections through the freedesktop standard via this interface. The downside is that the freedesktop standard is not a strong statement on APIs or compatibility-- it is little more than "here is a way an application can expose its dbus API to other apps" and as such this interface is a rather general purpose (but we can address abuse, see below).

"What happens if someone connects to that slot?"

Today, snaps cannot connect to the slot because there is no plugs side policy, but we can decide what that would look like as part of this PR or in a future iteration. Assuming that the yaml is worked out (see below for an idea), what I envision would happen is something like:

```
const dbusAppConnectedSlotAppArmor = `
# allow connected snaps to talk to this service's ###DBUS_BIND_NAME###
dbus (receive)
    bus=###DBUS_BIND_BUS###
    path=###DBUS_BIND_PATH###
    peer=(label=###PLUG_SECURITY_TAGS###),
`

const dbusAppConnectedPlugAppArmor = `
# allow connected snaps to talk to service implementing ###DBUS_BIND_NAME###
dbus (send)
    bus=###DBUS_BIND_BUS###
    path=###DBUS_BIND_PATH###
    peer=(label=###SLOT_SECURITY_TAGS###),
`
```

"How do we even hint to the user what can be connected to what?"

With this PR, the slot side might do:
```
name: rhythmbox
slots:
  rhythmbox-dbus:
    interface: dbus-app
    session:
    - org.gnome.Rhythmbox
```

With this we get:
```
$ snap interfaces
Slot                     Plug
rhythmbox:rhythmbox-dbus -
```

In the future iteration of this interface where we've defined how a snap might plug into a slotting snap, the plug side might do:
```
plugs:
  rhythmbox:
    interface: dbus-app
    session:
    - org.gnome.Rhythmbox
```

Then we get:
```
$ snap interfaces
Slot                     Plug
rhythmbox:rhythmbox-dbus -
-                        rhythmbox-remote:rhythmbox

$ snap connect rhythmbox-remote:rhythmbox rhythmbox:rhythmbox-dbus
$ snap interfaces
Slot                     Plug
rhythmbox:rhythmbox-dbus rhythmbox-remote:rhythmbox
```

"How do we prevent an app from binding?"

In today's snapd implementation, all snaps implementing slots are flagged for manual review in the store, but once it hits the device, snapd would unconditionally allow the bind since it is on the slot side.

Importantly, we've said that *if* a snap were to use this interface and bind to an address then that would be something that we'd review/approve in the store which means that once snap declarations are in place, snapd won't allow the snap to be installed unless it is approved. We would want store reviewers to review/approve regardless of if it is slot or plug side, so the install experience would be equivalent either way. What is different with this PR is that the user won't be able to do 'snap disconnect foo:dbus-app', but I think that is ok because the interface doesn't provide elevated access to the system, removing the access would make most apps crash in today's world, and we would auto-connect the interface if it was the plugs side since the approval of the well-known dbus name was approved in the store. All combined, there is little reason I can think of to give the user the ability to disconnect this interface. There is a lot of reason for a user to be able to disconnect rhythmbox (slot side) and rhythmbox-remote (plug side) though, and that would be supported.

"If the answers are respectively "nothing", "we don't know", and "we can't", then something is wrong."

As mentioned above, I feel that the current PR best reflects the intent of applications that bind to a well-known dbus name and allows for future growth where we connect plugging snaps. In my mind, the biggest drawback is that the interface is more general purpose than what we typically think of for interfaces, but we can ensure that snaps aren't abusing the interface when we review the snap in the store when they are requesting a well known dbus name. For example, 'org.gnome.Rhythmbox' or 'org.gnome.Evolution' we'd allow the use of this interface, but 'com.hp.hplip' or 'f1.w1.wpa_supplicant1' we would not.

"Taking our lesson from our conversation this week, isn't this actually more correct as dbus-support, where the settings are done on the plug side, and only when actually granted via a connection?"

I agree it should be named '<something>-support' if implementing on the plugs side. If we implement in the manner of this PR, we may want to change this to freedesktop-app instead.

zyga commented on interfaces/builtin/dbus_app.go:
```diff
@@ -0,0 +1,244 @@
+				return nil, fmt.Errorf("bus name is too long (must be <= 255)")
+			}
+
+			validBusName := regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9_-]*(\\.[a-zA-Z0-9_-]+)+$")
```
You could move this to module scope to reuse it

jdstrand replied:
Done. Added to interfaces/core.go

zyga commented on interfaces/builtin/dbus_app.go:
```diff
@@ -0,0 +1,244 @@
+	} else if bus == "session" {
+		abstraction = "dbus-session-strict"
+	} else {
+		return "", fmt.Errorf("unknown abstraction for specified bus '%s'", bus)
```
Use `%q` instead of `'%s'`

jdstrand replied:
Done

zyga commented on interfaces/builtin/dbus_app.go:
```diff
@@ -0,0 +1,244 @@
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus path
+	dot_re := regexp.MustCompile("\\.")
```
Another thing that could be reused if kept at the module level

jdstrand replied:
This seems very module specific. Ie, trying to convert `org.test-session1` to `/org/test-session1{,/**}`. This for writing a particular glob rule for policy and abstracting it out likely has little benefit and would make policy code reviews more difficult.

niemeyer commented:
@jdstrand Thanks for taking the time to so carefully explain the background. Perhaps you are right and the interface name is enough of a hint for people to connect dbus plugs to the right slots, and indeed this is a more powerful model than simply adding a support interface. So, as a change in direction following your suggestions, perhaps we should instead rename this to "dbus" (that's really the media between the apps), and try to sort out both the plug and the slots here to ensure we can support the scenario you describe. Then, with the nice snap-declaration support we discussed today, we should be able to moderate the usage of the bus very easily for both ends.

How does that sound?

niemeyer commented:
This is lingering and has open points for agreement. I'm closing it but please revive when it's been tweaked after whatever we agree to.

jdstrand commented:
Reopening since this should be in mergeable state again. This reopened PR doesn't change anything yet over what was proposed before. Ie:

```
name: rhythmbox
slots:
  rhythmbox-dbus:
    interface: dbus-app
    session:
    - org.gnome.Rhythmbox

$ snap interfaces
Slot                     Plug
rhythmbox:rhythmbox-dbus -
```

Or more generally:
```
name: foo
slots:
  bar:
    interface: dbus-app
    session:
    - com.example.baz
    - com.example.norf
    system:
    - com.example.qux

$ snap interfaces
Slot                     Plug
foo:bar                  -
```

Since you stated last time that we should handle the plugs side now, for the sake of discussion, what if we keep this very simple and let any snap plug into a slot implementation of dbus-app. For example for rhythmbox:
```
plugs:
  rhythmbox:
    interface: dbus-app

$ snap interfaces
Slot                     Plug
rhythmbox:rhythmbox-dbus -
-                        rhythmbox-remote:rhythmbox

$ snap connect rhythmbox-remote:rhythmbox rhythmbox:rhythmbox-dbus
$ snap interfaces
Slot                     Plug
rhythmbox:rhythmbox-dbus rhythmbox-remote:rhythmbox
```

Or more generally:
```
name: some-snap
plugs:
  foo:
    interface: dbus-app

$ snap interfaces
Slot                     Plug
foo:bar                  -
rhythmbox:rhythmbox-dbus -
-                        some-snap:foo

$ snap connect some-snap:foo foo:bar
$ snap interfaces
Slot                     Plug
foo:bar                  some-snap:foo
rhythmbox:rhythmbox-dbus -
```

One thing we want to discuss and make sure is covered in documentation is that there is nothing stopping the user from connecting some-snap to rhythmbox. Thus far, we've typically only seen one slot implementation per interface on a system at any one time. This general purpose interface is intended to allow multiple dbus-app slot implementations for the myriad of dbus apis out and there isn't way (in the proposed implementation) to restrict that further (ie, some-snap can be connected to both foo and rhythmbox instead of just foo). This is not a fatal condition though since while some-snap is then allowed to talk to rhythmbox, it isn't required to do so.

If we wanted to restrict thus further, we could add a required attribute on the slots and plugs side to more tightly couple the slot implementation to the plugging side. Eg:

```
name: rhythmbox
slots:
  interface: dbus-app
  api: rhythmbox
  session:
  - org.gnome.Rhythmbox

name: foo
slots:
  bar:
    interface: dbus-app
    api: foo
    session:
    - com.example.baz
    - com.example.norf
    system:
    - com.example.qux

name: some-snap
plugs:
  foo:
    interface: dbus-app
    api: foo
```

By calling it 'api', we can do things like 'foo2', 'foo3', etc to help with api breaks. We can of course call 'api' something different.

Alternatively, we could introduce some magic and say that the name of the interface reference should match on the plugs and slots side. I think this would be a mistake since it would require coordination between the slot snap and plug snap developers if the name should change where this is not the case currently.

Honestly, I'm not 100% sure to have something like 'api' or not. Having it adds a tighter coupling but perhaps that isn't desired for this interface....

jdstrand commented:
Removing 'Reviewed' tag since this is now ready to discuss again.

chipaca commented:
the test error should go away if you rebase

jdstrand commented:
@niemeyer - this interface is not needed for GA but it would be good to pick the discussion as soon as you are able. Please see https://github.com/snapcore/snapd/pull/1613#issuecomment-254050047

chipaca commented:
Is there a cheap way of making it OK to auto-connect, e.g. if api is given and matches?

jdstrand commented:
@chipaca - the way I envision this working based on past conversations (though it isn't expressed in the base declaration yet since @niemeyer and I haven't finalized the design yet) is that the slot implementation upload would specify the well-known bus names to bind to and this would prompt manual review. The reviewer can choose if those names are appropriate and create a snap declaration allowing it. That snap declaration could easily say 'allow-auto-connection: true'.

jdstrand commented:
@niemeyer and I discussed this today and decided on a different approach for the yaml and renaming this simply `dbus`.

Slot implementations would offer DBus interfaces on particular buses, like so:

```
name: foo
slots:
  my-dbus-interface1:
    interface: dbus
      name: org.my.dbus.Interface1
      bus: session
  my-dbus-interface2:
    interface: dbus
      name: org.my.dbus.Interface2
      bus: system
```

Consumers would plug the interfaces they care about. Eg:

```
name: bar
plugs:
  foo-dbus-interface1:
    interface: dbus
      name: org.my.dbus.Interface1
      bus: session
```

Connections look like:
```
$ snap interfaces
Slot                     Plug
foo:my-dbus-interface1   -
foo:my-dbus-interface2   -
-                        bar:foo-dbus-interface1

$ snap connect bar:foo-dbus-interface1 foo:my-dbus-interface1
$ snap interfaces
foo:my-dbus-interface1   bar:foo-dbus-interface1
foo:my-dbus-interface2   -
```

AppArmor security policy for `foo` has something like:
```
dbus (bind)
  bus=session
  name=org.my.dbus.Interface1,

dbus (bind)
  bus=system
  name=org.my.dbus.Interface2,

dbus (receive, send)
  bus=session
  interface=org.my.dbus.Interface1{,.*},
  peer=(label=snap.bar.bar),
```


and bar has:
```
dbus
  bus=session
  interface=org.my.dbus.Interface1{,.*},
  peer=(label=snap.foo.foo),
```

This solves the problem of GNOME and KDE (or any freedesktop.org) apps wanting to bind to a well-known name. This allows KDE applications to connect to one another via DBus by simply declaring what dbus well-known name and buses they are interested in and ultimately fixes LP #1590679.

This also sets the stage for the future once we learn more how people want to use this interface since it could allow for simplification of future interfaces where an entire DBus API should be allowed upon connection (ie, for session services). We can continue to use the current methodology of specific interfaces when carving up DBus APIs (eg, location-observe vs location-control) and when needing to specify dbus bus policy (ie, for system services using the dbus backend in snapd), but can continue thinking about how to extend this new 'dbus' interface for these use cases (ie, bus policy, DBus 'path' and 'member').

jdstrand commented:
Removing Blocked tagged and closing for now. I'll re-open when have something concrete to review.

tyhicks commented:
@jdstrand The example rule you gave for bar's plug doesn't explicitly specify the allowed permissions. Please list "(receive, send)" so there's no confusion and we're future-proofed against new dbus permissions being added in the future which were not meant to be granted.

jdstrand commented:
@tyhicks - fixed, thanks (simple omission on my end).

jdstrand commented:
FYI, this is implemented here: https://github.com/snapcore/snapd/compare/master...jdstrand:dbus-app?expand=1 but I want to do more testing before reproposing the PR.

jdstrand commented:
FYI this was tested with gnome-logs (for the bind) and with d-feet (unconfined and running under snap-equivalent confinement). I asked for representative KDE apps for interface connections but haven't heard back yet (we can adjust policy as necessary once KDE folks use the interface-- the gnome-logs and d-feet connection is sufficient to show the feature works).

jdstrand commented:
Ok, I heard back from KDE folks. Here is the test case:
$ sudo snap install kde-frameworks-5 --edge
$ sudo snap install ktuberling --edge
$ sudo snap connect ktuberling:kde-frameworks-5-plug kde-frameworks-5:kde-frameworks-5-slot

May have to (eg, if tried to launch without kde-frameworks-5):
$ sudo /usr/lib/snapd/snap-discard-ns ktuberling

At this point it shows a denial. Unfortunately, the denial is:
apparmor="DENIED" operation="dbus_bind"  bus="session" name="org.kde.ktuberling-7881" mask="bind" pid=7881 label="snap.ktuberling.ktuberling"

so it is using its PID as part of the 'well-known name'. :(

niemeyer commented:
@jdstrand How can we support the pid?  Can apparmor deal with wildcards?

niemeyer commented:
Well, even if it can, that won't quite work as intended, as we won't be able to match plug and slot.

I don't understand how this is being used.. how can it use a pid as part of the interface?  What is the dbus client doing to talk to this server object?

This misunderstanding may be fundamental to the overall feature, so let's make sure to clear this up before we move on with this plan.

jdstrand commented:
@niemeyer - ktuberling is abusing the dbus bind and only having part of the name be well-known. I'm guessing the reason why they are doing this is so that multiple instances of ktuberling can be launched within the same session and that KDE apps will register themselves with some KDE service and other KDE apps will use that service to find the various ktuberlings to access them. GNOME doesn't do this and is compliant with the freedesktop.org spec: https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names

I don't think this a misunderstanding of the overall feature, but we have to think about how to support it for KDE. We can support the pid in the security policy by simply adding an alternation to the rule. Eg, slot still has 'name: org.kde.ktuberling' but we create a bind rule for it like so:
```
dbus (bind)
    bus=session
    name=org.kde.ktuberling{,-[1-9],-[1-9][0-9],-[1-9][0-9][0-9],-[1-9][0-9][0-9][0-9],-[1-9][0-9][0-9][0-9][0-9],-[1-9][0-9][0-9][0-9][0-9][0-9]},
```

While it is somewhat unfortunate the way that KDE apps are misusing the well-known name, we can deal with it. There is a problem with the above in that if KDE creates 'ktuberling' and I create 'ktuberling-12345' then do to the above rules, we have an overlap:

```
snap             PID   allowed
------------------------------
ktuberling       12345 org.kde.ktuberling,org.kde.ktuberling-1,...,org.kde.ktuberling-12345
ktuberling-12345 ...   org.kde.ktuberling-12345,org.kde.ktuberling-12345-1,...
```

This could be addressed in a few ways:
- we ignore the overlap
- we don't allow the name attribute to end with `-[0-9]+`
- we don't worry about it in snapd and handle it in the store

jdstrand commented:
I suggest for the time being we don't allow the name attribute to end with -[0-9]+ in snapd and add the alternation unconditionally then see how real people use the interface.

jdstrand commented:
FYI, the introspection rule needs to be adjusted for KDE, but that is less of a concern.

jdstrand commented:
In the interest of time, I've added an alternation rule that allows KDE snaps to bind to org.kde.foo-12345 when they simply declare 'name: org.kde.foo'. I also adjusted ValidateDBusBusName() in interfaces/core.go to not allow bus names ending with '-[0-9]+'. This changes can be easily reverted if needed or iterated on in the future.

niemeyer commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,381 @@
+	}
+
+	// ensure that we only connect to slot with matching attributes
+	if bus != busPlug || name != namePlug {
```
This should be an error. We should not even allow that connection to be established.

jdstrand replied:
I initially had this as an error but throwing an error here means that situations like TestConnectionFirst() and TestConnectionSecond() in dbus_test.go throw an error, but they should not. I'm trying to discard 'the other' one so just skip it. Is there a better way?

niemeyer replied:
Why they should not?  I feel like I'm missing something important, perhaps unrelated to this PR. If we're running the connection code for things which are incompatible and that should not be auto-connected, erroring seems sane. If it's blowing up in a place it shouldn't, definitely curious to hear about it.

jdstrand replied:
I haven't looked at this terribly closely, but it looks like this is cause by ResolveConnect() in interfaces/repo.go:

```
func (r *Repository) ResolveConnect(plugSnapName, plugName, slotSnapName, slotName string) (ConnRef, error) {
...
        if slotName == "" {                                                     
                // Find the unambiguous slot that satisfies plug requirements   
                var candidates []string                                         
                for candidateSlotName, candidateSlot := range r.slots[slotSnapName] {
                        // TODO: use some smarter matching (e.g. against $attrs)
                        if candidateSlot.Interface == plug.Interface {          
                                candidates = append(candidates, candidateSlotName)
                        }                                                       
                }
```

Note `// TODO: use some smarter matching (e.g. against $attrs)` - it isn't matching against attrs so the interface code must when matching like this. This makes sense with my observations that I could not error and instead must silently ignore attribute mismatches since Connected*Snippet  was getting each combination.

niemeyer replied:
Okay, we need to look into the auto-connection rule based on attributes soon anyway, so let's not block on this.

pedronis replied:
yes, it looks like that code in repo.go should use the policy checks, especially once we have attribute comparision

pedronis replied:
@jdstrand could you replace that TODO there with  use interface/policy Checkers

jdstrand replied:
Added a TODO comment

niemeyer commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,381 @@
+}
+
+// Since we only implement the permanent slot side, this is meaningless but
+// we have to supply the method, so set it to something safe.
```
"true" doesn't seem very safe. :)

jdstrand commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,402 @@
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus path (eg 'org.foo' to '/org/foo{,/**}')
+	dot_re := regexp.MustCompile("\\.")
```
I found that I couldn't get rid of dbus path entirely due to snaps using the org.freedesktop.* interfaces and so instead have:

```
139 # allow connected snaps to ###DBUS_NAME###                                      
140 dbus (receive, send)
141     bus=###DBUS_BUS###
142     interface=###DBUS_INTERFACE###
143     peer=(label=###SLOT_SECURITY_TAGS###),
144                                                                                 
145 # allow connected snaps to all interfaces via ###DBUS_PATH### (eg,              
146 # org.freedesktop.*, org.gtk.Application, etc) to allow full integration with   
147 # connected snaps.                                                              
dbus (receive, send)
    bus=###DBUS_BUS###
    path=###DBUS_PATH###
    peer=(label=###SLOT_SECURITY_TAGS###),  
```

This allows us to match by either interface (and security peer label) or path (and security peer label) and avoids the problems of trying to specify them both at the same time. In other words, still have the rule you mention that doesn't require guessing, but then have another rule that allows using the org.freedesktop.DBus interface in predictable ways.

 I discussed this with @tyhicks and we agreed on the approach.

niemeyer replied:
That still doesn't look right. This is an interface name, not a path. If we are translating from one to the other, we're doing guess work which can be wrong, because after all those are two different fields in the dbus system instead of a single one. If the interface matching doesn't work for some cases which require a path instead of an interface, let's please raise that and talk about it in those terms.

niemeyer replied:
Okay, sounds fine. We can always include an additional `path` field if we need to be explicit.

jdstrand commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,381 @@
+	new = pathBuf.Bytes()
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus interface (eg, 'org.foo' to 'org.foo{,.*}')
```
No, it is accurate. It takes 'name' which might be 'org.foo' and turns it into the glob 'org.foo{,.*}'.

jdstrand commented on interfaces/builtin/basedeclaration.go:
```diff
@@ -203,6 +203,14 @@ slots:
+    deny-connection:
+      slot-attributes:
+        name: .+
+    deny-auto-connection: true
```
Yes. I thought we agreed that people would need to have their well-known dbus name approved in the store. This will trigger a manual review and enforce the use of a snap declaration to claim the well-known name.

jdstrand replied:
Also, it doesn't prevent them from using the interface-- they can install a snap that slots this interface and get the bind rules needed to run; it is just that you can't connect other snaps to this interface until a snap declaration is present.

I think this is the right approach for this interface and others in general, but it sheds a light on a developer pain point as described in https://bugs.launchpad.net/snappy/+bug/1640874. I think this is perhaps solved by a locally signed snap declaration, but that needs your input (see my comments in the bug).

jdstrand replied:
We reject it because the snap has not claimed the name via the store and doesn't yet have a snap declaration. On initial upload with using this interface, the reviewer will issue a snap declaration that allows the connection, and then all is well for connections and subsequent uploads. I think this is in line with everything we've discussed regarding the base declaration and snap declarations-- slot implementation (typically) need some sort of snap declaration since it (typically) allows privileged access to the system. In this case the privilege is to claim a well-known name.

On a related note, the review tools are now considering the base declaration and prompting for manual review due to this constraint. If we remove the constraint today, the tools would let this through without prompting for manual review. If we decide that the constraint should not be there, then the tools will have to be modified to prompt for manual review based on factors outside of the base declaration (which is totally doable, but I thought one goal of the base declaration was to be the one place the review tools would consult for how interface reviews were performed).

jdstrand commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,402 @@
+func (iface *DbusInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
+	bus := ""
+	name := ""
+	for attr := range attribs {
```
Fixed

jdstrand commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,418 @@
+<policy context="default">
+    <deny send_destination="###DBUS_NAME###"/>
+</policy>
+`
```
dbusPermanentSlotDBus wasn't in the original PR but it is needed for 'bus=system'. @niemeyer and @tyhicks: the question then becomes, should non-root be allowed to talk to it? As implemented (in a separate commit for easy reverting), the answer is 'yes' because snapd should allow a non-root snap command should be able to connect to a root daemon within the same command and it is up to the snap to perform any uid checks. However, because of the dbusPermanentSlotAppArmorClassic policy for on classic (and the lack of policykit in the core snap) this means that unconfined non-root can talk to the snap-- this needs discussion because this interface does alone not grant extra access to the system, b) this interface in combination with privileged ones (eg, network-control) requires a user to connect the interfaces and c) unconfined non-root can aa_change_profile() to something that would allow the access. 

'b' is somewhat mediated in the initial store upload since the base declaration requires that to use this interface you need to claim the name and receive a snap declaration. However, the tools aren't considering combinations of interfaces and it would be a shame if a system, root-running slot implementation service has say, network-control and exposes a DBus API but doesn't perform uid checks so non-root users could escalate privileges to change networking via the poorly implemented snap. Granted, there is nothing saying this poorly written snap couldn't expose a named unix socket in SNAP_DATA world read/write and to the same thing without this interface, so this is probably one of those things where this poorly written gets poor reviews and/or bug report and we exercise store policies to compel the developer to fix it (or revoke the snap).

@niemeyer and @tyhicks, should non-root be able to contact a dbus system service via this interface?

jdstrand commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,418 @@
+<policy context="default">
+    <allow send_destination="###DBUS_NAME###"/>
+</policy>
+`
```
dbusPermanentSlotDBus wasn't in the original PR but it is needed for 'bus=system'. @niemeyer and @tyhicks: the question then becomes, should non-root be allowed to talk to it? As implemented (in a separate commit for easy reverting), the answer is 'yes' because snapd should allow a non-root snap command should be able to connect to a root daemon within the same command and it is up to the snap to perform any uid checks. However, because of the dbusPermanentSlotAppArmorClassic policy for on classic (and the lack of policykit in the core snap) this means that unconfined non-root can talk to the snap-- this needs discussion because this interface does alone not grant extra access to the system, b) this interface in combination with privileged ones (eg, network-control) requires a user to connect the interfaces and c) unconfined non-root can aa_change_profile() to something that would allow the access. 

'b' is somewhat mediated in the initial store upload since the base declaration requires that to use this interface you need to claim the name and receive a snap declaration. However, the tools aren't considering combinations of interfaces and it would be a shame if a system, root-running slot implementation service has say, network-control and exposes a DBus API but doesn't perform uid checks so non-root users could escalate privileges to change networking via the poorly implemented snap. Granted, there is nothing saying this poorly written snap couldn't expose a named unix socket in SNAP_DATA world read/write and to the same thing without this interface, so this is probably one of those things where this poorly written gets poor reviews and/or bug report and we exercise store policies to compel the developer to fix it (or revoke the snap).

@niemeyer and @tyhicks, should non-root be able to contact a dbus system service via this interface?

(sorry for the second email on this, I made a change and the comment was hidden so I deleted the first and added this one)

jdstrand commented:
The trusty issue is unrelated to this PR:
Exit request sent.
blame: https://github.com/snapcore/snapd.git#refs/pull/1613/head
badpkg: Test dependencies are unsatisfiable. A common reason is that your testbed is out of date with respect to the archive, and you need to use a current testbed or run apt-get update or use -U.
autopkgtest [21:00:56]: ERROR: erroneous package: Test dependencies are unsatisfiable. A common reason is that your testbed is out of date with respect to the archive, and you need to use a current testbed or run apt-get update or use -U.

jdstrand commented:
Added the 2.20 milestone since AIUI all this needs is one more review and this interface has been lingering. @pedronis - AIUI you will be doing that review, is that correct?

pedronis commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,430 @@
+	dot_re := regexp.MustCompile("\\.")
+	var pathBuf bytes.Buffer
+	pathBuf.WriteString(`"/`)
+	pathBuf.WriteString(dot_re.ReplaceAllString(name, "/"))
```
the bit stands that this could be just strings.Replace(name, ".", "/", -1)

jdstrand replied:
Done

pedronis commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,430 @@
+}
+
+// Obtain yaml-specified bus well-known name
+func (iface *DbusInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
```
would be better to use named results here: (bus, name string, err error)

pedronis commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,430 @@
+
+func (iface *DbusInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) bool {
+	// allow what declarations allowed
+	return true
```
the check // ensure that we only connect to slot with matching attributes
 +	if bus != busPlug || name != namePlug {

should probably be done here? instead?

that's what content is doing atm btw, until we don't have support for specifying attribute matching in the basedecl

jdstrand replied:
AutoConnect() isn't the right place for the check for the dbus interface because the check needs to happen at connection, not auto connection and at connection we only connect when bus and name match. The content interface is different in that while the content attribute must match for auto-connect, the content attribute is not used as part of the connected snippets, only read and write which aren't part of the content interface's plugging attributes.

Note that the dbus interface design and its attributes are approved.

pedronis commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,430 @@
+		return "", "", fmt.Errorf("DBus bus name must not end with -NUMBER")
+	}
+
+	if bus == "" {
```
I'm confused by these checks, it doesn't seem that bus == "" would pass the previous check being equal session or system,  same for name == "" and ValidateDBusName ?

jdstrand replied:
@pedronis, you are right. These are a holdover from a refactor. Removed.

jdstrand commented:
@pedronis, I believe I addressed your comments.

jdstrand commented:
@pedronis and I discussed the 'if bus != busPlug || name != namePlug' part on irc and agreed that what is implemented is the best we can do without the attribute matching. The code will ensure that the correct security policy is generated. I updated the comment to reflect this.

jdstrand commented:
@pedronis and @tyhicks - I believe I addressed all your comments. Can you give a +1 or more feedback if needed?

Thanks for your reviews!

tyhicks commented:
The security policy looks good to me. Thanks for making the adjustments we spoke about over IRC.

jdstrand commented:
@niemeyer - I believe this is mergeable now based on my incorporating feedback from you, @pedronis, and @tyhicks. Please consider approving the changes and pending the testsuite run.

jdstrand commented:
The travis testsuite failure is unrelated:
2016/12/14 00:37:47 Failed tasks: 1
    - linode:ubuntu-core-16-64:tests/main/revert-devmode:production

+ snap refresh --devmode --edge test-snapd-tools
error: cannot refresh "test-snapd-tools": cannot get refresh information for snap "test-snapd-tools": Post https://search.apps.ubuntu.com/api/v1/snaps/metadata: EOF

jdstrand commented:
The trusty failure is unrelated:
Exit request sent.
autopkgtest [01:12:08]: ERROR: erroneous package: Test dependencies are unsatisfiable. A common reason is that your testbed is out of date with respect to the archive, and you need to use a current testbed or run apt-get update or use -U.
blame: https://github.com/snapcore/snapd.git#refs/pull/1613/head
badpkg: Test dependencies are unsatisfiable. A common reason is that your testbed is out of date with respect to the archive, and you need to use a current testbed or run apt-get update or use -U.

jdstrand commented:
The zesty failures are unrelated:
```
2016/12/14 01:22:14 Error executing autopkgtest:ubuntu-16.04-64:tests/main/server-snap:pythonServer : 
-----
++ nc -4 localhost 80
+ response=
+ statusPattern='(?s)HTTP\/1\.0 200 OK\n*'
+ grep -Pzq '(?s)HTTP\/1\.0 200 OK\n*'
+ echo ''
```

```
2016/12/14 01:11:22 Error executing autopkgtest:ubuntu-16.04-64:tests/main/interfaces-network-bind : 
-----
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/names.sh
+++ sed -n 's/^\(pc\|pi[23]\|dragonboard\) .*/\1/p'
+++ snap list
++ gadget_name=
++ kernel_name=-kernel
+++ awk '/^(ubuntu-)?core / {print $1; exit}'
+++ snap list
++ core_name=core
++ '[' -kernel = pi3-kernel ']'
+ CONNECTED_PATTERN='(?s)Slot +Plug\n.*?\n:network-bind +network-bind-consumer'
+ DISCONNECTED_PATTERN='(?s)Slot +Plug\n.*?\n- +network-bind-consumer:network-bind'
+ echo 'Then the snap is listed as connected'
Then the snap is listed as connected
+ grep -Pzq '(?s)Slot +Plug\n.*?\n:network-bind +network-bind-consumer'
+ snap interfaces
+ echo ============================================
============================================
+ echo 'When the plug is disconnected'
When the plug is disconnected
+ snap disconnect network-bind-consumer:network-bind core:network-bind

[|] Disconnect network-bind-consumer:network-bind from core:network-bind[K
[K+ grep -Pzq '(?s)Slot +Plug\n.*?\n- +network-bind-consumer:network-bind'
+ snap interfaces
+ echo 'Then the plug can be connected again'
Then the plug can be connected again
+ snap connect network-bind-consumer:network-bind core:network-bind

[|] Connect network-bind-consumer:network-bind to core:network-bind[K
[K+ grep -Pzq '(?s)Slot +Plug\n.*?\n:network-bind +network-bind-consumer'
+ snap interfaces
+ echo ============================================
============================================
+ echo 'When the plug is connected'
When the plug is connected
+ snap connect network-bind-consumer:network-bind core:network-bind

[|] Connect network-bind-consumer:network-bind to core:network-bind[K
[K+ grep -Pzq '(?s)Slot +Plug\n.*?\n:network-bind +network-bind-consumer'
+ snap interfaces
+ echo 'Then the service is accessible by a client'
Then the service is accessible by a client
+ grep -Pqz 'ok\n'
+ nc -w 2 localhost 8081
```

```
2016/12/14 01:22:18 Error executing autopkgtest:ubuntu-16.04-64:tests/main/server-snap:goServer : 
-----
++ nc -6 ip6-localhost 8081
+ response=
+ statusPattern='(?s)HTTP\/1\.0 200 OK\n*'
+ grep -Pzq '(?s)HTTP\/1\.0 200 OK\n*'
+ echo ''
```

jdstrand commented:
Retrying the travis tests since there seem to be intermittent failures.

jhenstridge commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,426 @@
+    bus=###DBUS_BUS###
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="GetConnectionUnix{ProcessID,User}"
```
Would it make sense to also include org.freedesktop.DBus.GetConnectionCredentials here?  That seems to be the replacement for the above two if you need both.

jhenstridge commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,426 @@
+	var ifaceBuf bytes.Buffer
+	ifaceBuf.WriteString(`"`)
+	ifaceBuf.WriteString(name)
+	ifaceBuf.WriteString(`{,.*}"`)
```
This might make sense as a default value, but interface names are not the same as bus names in the general case.  For my work on storage-framework, we have a number of daemons that all implement the same D-Bus interface, but will register on the bus with different well known names.

I believe MPRIS works similarly.

Also, what if the service exposes method calls or signals on more than one interface? I imagine it'd be quite common for an app to want to respond to method calls on the org.freedesktop.DBus.Properties interface, but I think it'd be a mistake to simply special case it.

jdstrand replied:
As mentioned on the mailing list: https://lists.ubuntu.com/archives/snapcraft/2016-December/001998.html this initial implementation of the interface intentionally does not address complex dbus services. It is meant to address leaf applications (eg, GNOME and KDE) that bind to a well-known name.

I'd prefer not to delay this PR any longer by adding more functionality to it. @niemeyer and I agreed in this PR that we would implement what we have and then see how people are using it and potentially add DBus paths, interfaces, activation, complex bus policy, etc down the line. Due keep in mind that generic interfaces like this one are a new concept in snappy that is in some ways counter to the initial goals of interfaces, and we want to make sure we design and implement them in a controlled fashion.

As for storage framework in particular, if you haven't already, I suggest reading the aforementioned thread. At this time, it will need to be its own interface (and that's ok-- it's easy getting interfaces into snappy :). Perhaps at a future date the dbus interface will get to the point to support all DBus services generically, but not just yet.

jhenstridge replied:
If you don't want to make it configurable, then I'd strongly consider allowing at least org.freedesktop.DBus.Properties as an extra one unconditionally.  Without it, it is impossible to access APIs using the standard D-Bus property system.

To get property "A" on interface "I", you call org.freedesktop.DBus.Properties.Get("I", "A") -- not a method call on the interface declaring the property.

jdstrand replied:
To be clear, we do want to make it configurable-- at some point. There is an urgency to this PR as it is blocking GNOME and KDE devs from targeting stable. This PR specifically addresses that use case. Other use cases can come in other PRs.

As for Properties, this already handled by the path rule without an interface:
```
# allow connected snaps to all interfaces via ###DBUS_PATH### (eg,
# org.freedesktop.*, org.gtk.Application, etc) to allow full integration with
# connected snaps.
dbus (receive, send)
    bus=###DBUS_BUS###
    path=###DBUS_PATH###
    peer=(label=###SLOT_SECURITY_TAGS###),
```

This was a tested condition for the leaf applications we are trying to address.

jhenstridge commented on interfaces/builtin/dbus.go:
```diff
@@ -0,0 +1,426 @@
+	var pathBuf bytes.Buffer
+	pathBuf.WriteString(`"/`)
+	pathBuf.WriteString(strings.Replace(name, ".", "/", -1))
+	pathBuf.WriteString(`{,/**}"`)
```
Similar to the comment about interface names, you can't necessarily derive the paths a D-Bus service uses from the bus name it will try to acquire.  If you're going to restrict object paths (which seems a bit dubious), shouldn't it be configurable on the slot?

jdstrand commented:
Travis failure is unrelated: linode:ubuntu-core-16-64-fixme:tests/main/ubuntu-core-upgrade-no-gc

Trusty autopkgtest failure is unrelated

Zesty autopkgtest failures are unrelated (and same as before):
    - autopkgtest:ubuntu-16.04-64:tests/main/interfaces-network-bind
    - autopkgtest:ubuntu-16.04-64:tests/main/server-snap:goServer
    - autopkgtest:ubuntu-16.04-64:tests/main/server-snap:pythonServer

jdstrand commented:
From IRC:
```
12:58 < niemeyer> So do we support any interface on the well known name?
13:01 < jdstrand> we allow a connected snap to introspect our dbus api
13:01 < jdstrand> we allow a connected snap to use any dbus interface and any dbus 
                  path if they are connecting to the well-known name
13:02 < jdstrand> if they are using a non-well-known name, we allow connecting to any 
                  dbus path if the dbus interface is regular according to our 
                  well-known name
13:02 < niemeyer> Okay, that sounds reasonable
13:02 < niemeyer> Thanks for the long explanation
13:02 < jdstrand> if they are using a non-well-known name, we allow connecting with 
                  any dbus interface if the dbus path is regular accoriding to our 
                  well-known name
13:03 < niemeyer> Happy to go forward with it
13:03 < jdstrand> \o/
13:03 < niemeyer> Is the PR blocked on anything else at this point?
13:04 < jdstrand> niemeyer: no. tyhicks reviewed the security policy. pedronis 
                  reviewed the code. I incorporated all applicable feedback from 
                  everywhere
13:04 < niemeyer> So let's get it in
13:04 < jdstrand> thanks!
13:04 < niemeyer> Thank you!
13:06 < jdstrand> niemeyer: so with that, I plan to work with the personal folks on 
                  designing their interfaces (eg, storage framework) and will be 
                  reviewing tvoss' work for session services and will be thinking 
                  about how all of this relates to this interface, improvements we 
                  can make to this interface, etc and when I have coherent thoughts 
                  on it, will present phase ii, phase iii, etc, etc iterations to you
```

jdstrand commented:
Thanks especially @niemeyer for your time on this and thank you to all reviewers and commenters for your input and reviews! :)
